roofit/roofitcore/inc/SpanCast.h

```diff
@@ -0,0 +1,111 @@
+// Author: Jonas Rembser, CERN, Jul 2021
+
+#ifndef roofit_roofitcore_SpanCast_h
+#define roofit_roofitcore_SpanCast_h
+
+namespace detail {
+
+template <bool isDynamic, typename T>
+struct TemplatedCaster {
+   template <typename U>
+   static T cast(U const &u);
+};
+
+template <typename T>
+struct TemplatedCaster<false, T> {
+   template <typename U>
+   static T cast(U const &u)
+   {
+      return static_cast<T>(u);
+   }
+};
+
+template <typename T>
+struct TemplatedCaster<true, T> {
+   template <typename U>
+   static T cast(U const &u)
+   {
+      return dynamic_cast<T>(u);
+   }
+};
```

Comment:
I don't this this solves the problem of deciding on `dynamic_cast` vs `static_cast` at compile time, no? In C++ it would be so easy with `if constexpr` ... I added some more comments to explain why the boilerplate code is needed.

Reply:
> I don't this this solves the problem of deciding on dynamic_cast vs static_cast at compile time,

yes it should (although constexpr-if is best): https://godbolt.org/z/dj3zshjqW

If you anyway prefer structs instead of functions+tag types, you can reduce the boilerplate a bit by declaring only one specialization instead of 2:

```cpp
template <bool isDynamic = false, typename T>
struct TemplatedCaster {
   static T cast(U const &u)
   {
      return static_cast<T>(u);
   }
};

template <typename T>
struct TemplatedCaster<true, T> {
   template <typename U>
   static T cast(U const &u)
   {
      return dynamic_cast<T>(u);
   }
};
```

Reply:
My take in this:
https://godbolt.org/z/18ndvr35x